### DISCUSSION TITLE
Proposal: Inline Reference Type Construction

### DISCUSSION THREAD

**Discussion by mkrebser** [dotnet/runtime#84589] [General]
It would be convenient if we could determine where reference types are allocated. The motivation is to reduce the amount of indirection in complex objects that are composed of nested heap-allocated reference types.

Example 1:
```
using customLibrary.ReferenceTypes.That.I.DontWantTo.ReimplementAsStruct;

namespace inlineStuff;

public class ExampleClass {
        // Inline<T> is a special struct provided by .NET
	private Inline<object> _mutex;
	private Inline<CustomObject> _objectOne;
	private Inline<CustomObject> _objectTwo;
	private Inline<Dictionary<String, CustomObject2>> _objectMap;
}
```
In the first example, I have marked reference types as `inline` which will instruct the runtime to allocate the `Inline<refTypes>` and their ValueType fields in a contiguous tightly packed block- which together will be the ExampleClass. 

The goal here, being reducing amount of indirection of accessing all the different reference types which are normally allocated individually to the heap.

There are some restrictions here of course. For example:
 1. All of the inline objects must have a default constructor which is called when the ExampleClass is constructed.
 2. The inline objects are only ever freed when the root object (ExampleClass) is freed. 
 3. `struct Inline<T>` has an overloaded assignment operator that only accepts objects who implement `ICopyable<T>`. `ICopyable` is an copy interface which has the method `void copy<T>(in T from)`
    - Note* `T` does not _have to implement ICopyable_ to be used with `Inline<T>`. This generic constraint only applies to the assignment operator and `copy<T>(in T from)` function.
    - Note* It is not possible to overload assignment operator, so the runtime/compiler would need to do something special here
 5. `struct Inline<T>` has a `T getRef()` which just returns a normal reference type of T! (not a ref T though). The reference type returned by `getRef` can act like any normal reference type!
     -  It wouldn't technically be a 'perfectly normal' reference type. An Inline reference would prevent garbage collection of the root object- and is never collected on its own! (Collecting a root object collects all the inline objects it is composed of)
 6. Member variables which are array types cannot be inline. ie, cannot have `Inline<object[]>`
 7. There cannot be infinitely recursive in-lining.
 8. Polymorphism doesn't work on Inline member variables- they cannot be assigned, so an `Inline<BaseT>` is always exclusively an `Inline<BaseT>` (The Copyable:copy<T> interface can still be used for copying into them, though)

Example 2:
```
namespace inlineStuff;

public class Program {
	public static void Main() {
		var myArray = new Inline<ExampleClass>[20];
	}
}
```
In Example 2, I have allocated 20 reference types (from example 1) contiguously to an array. All of the same conditions from example 1 apply here too.. The array elements will not be freed from memory until the root (the array object) is freed. 

So I know what you're thinking... why not use structs? And this is true! _In my own code I can use structs._  However, most of the code I use is not mine! (like the .NET library) and has many reference types which I would love to use more efficiently.

If we don't like the new `Inline<T>` wrapper, then there could also just be a new keyword `inline`.
`private inline object _myObject;` `var myArray = new inline ExampleClass[20];`
(I think a Inline class would be easier to implement though, albeit requiring some special runtime & compiler updates)

**Comment by CyrusNajmabadi** [comment]
Moving to runtime as this doesn't appear to be a language request.  Up to runtime team if htey want to expose this special `Inline<T>` system.  If they did, we could consider some sort of special lang support (like a new keyword) if we really wanted it to be a core part of C#.

**Comment by kevingosse** [comment]
One issue with this is inheritance. 

Say you have:

```
public class Parent : ICopyable
{
    public long Field1;
}

public class Child : Parent
{
    public long Field2;
}
```

`Parent` is 8 bytes, `Child` is 16 bytes. So how much memory should the runtime reserve in this case?

```
public class ExampleClass 
{
    private Inline<Parent> _instance;

    public void SetInstance(Parent instance)
    {
        _instance = instance;
    }
}
```

If it reserves 8 bytes, then when trying to store an instance of `Child` it will be truncated (an issue known as object slicing in C++). That's the reason why inheritance is not allowed on structs.

**Comment by mkrebser** [comment]
@kevingosse Inheritance actually works fine! And by works fine.. I mean it doesn't work on Inline<T> in cases where 'T' has child classes.

Lets examine what happens!

Say I have some base class: `Inline<MyBaseClass> _field1;`

The default constructor is always used to initialize the object- and enough memory is allocated only for `MyBaseClass`. This is the same thing that happens when you call `new MyBaseClass()`!

Now lets say I want to do this! case1: `this._field1 = new Inline<MyChildClass>()` or cases2: `this._field1.getRef() = new MyChildClass()`

1. In case1: the `ICopyable:copy<T>` function is called! Remember, `struct Inline<T>` _has an assignment operator that will call ICopyable:copy on the MyBaseClass field instance_ So in this case, the base class implementation of copy<T> is invoked and the copy will be successfully performed!
2. And case2 just results in a compiler error because getRef() is a result from a function and cannot be assigned!


I have updated the post to specify that Inline<T> member variables are generally expected to NOT be polymorphic, hopefully this clears any confusion

**Reply by kevingosse** [reply]
> Remember, struct Inline<T> has an assignment operator that will call ICopyable:copy on the MyBaseClass field instance

Oh I misunderstood that bit. I thought Copy was Copy**To** and not Copy**From**. Indeed, it solves the problem I raised 👍 

As a side-note:

> So I know what you're thinking... why not use structs? And this is true! In my own code I can use structs. However, most of the code I use is not mine! (like the .NET library) and has many reference types which I would love to use more efficiently.

You would only be able to use it on types that implement `ICopyable`, so you still depend on the goodwill of the library maintainer.

**Reply by mkrebser** [reply]
@kevingosse I have updated the post to be more clear that this generic constraint only applies to the `void copy<T>(in T from) where T: ICopyable` method and is not a generic constraint on the `Inline<T>` struct definition
